llama_stack/apis/memory_banks/memory_banks.py

```diff
@@ -20,59 +22,121 @@ class MemoryBankType(Enum):
     graph = "graph"
 
 
-class CommonDef(BaseModel):
-    identifier: str
-    # Hack: move this out later
-    provider_id: str = ""
+@json_schema_type
+class MemoryBank(Resource):
+    type: Literal[ResourceType.memory_bank.value] = ResourceType.memory_bank.value
+    memory_bank_type: MemoryBankType
+
+
+@json_schema_type
+class VectorMemoryBank(MemoryBank):
+    memory_bank_type: Literal[MemoryBankType.vector.value] = MemoryBankType.vector.value
+    embedding_model: str
+    chunk_size_in_tokens: int
+    overlap_size_in_tokens: Optional[int] = None
+
+
+@json_schema_type
+class KeyValueMemoryBank(MemoryBank):
+    memory_bank_type: Literal[MemoryBankType.keyvalue.value] = (
+        MemoryBankType.keyvalue.value
+    )
+
+
+@json_schema_type
+class KeywordMemoryBank(MemoryBank):
+    memory_bank_type: Literal[MemoryBankType.keyword.value] = (
+        MemoryBankType.keyword.value
+    )
+
+
+@json_schema_type
+class GraphMemoryBank(MemoryBank):
+    memory_bank_type: Literal[MemoryBankType.graph.value] = MemoryBankType.graph.value
+
+
+@json_schema_type
+class BaseRegistration(BaseModel):
+    memory_bank_id: str
+    provider_resource_id: Optional[str] = None
+    provider_id: Optional[str] = None
 
 
 @json_schema_type
-class VectorMemoryBankDef(CommonDef):
-    type: Literal[MemoryBankType.vector.value] = MemoryBankType.vector.value
+class VectorRegistration(BaseRegistration):
     embedding_model: str
     chunk_size_in_tokens: int
     overlap_size_in_tokens: Optional[int] = None
 
 
 @json_schema_type
-class KeyValueMemoryBankDef(CommonDef):
-    type: Literal[MemoryBankType.keyvalue.value] = MemoryBankType.keyvalue.value
+class KeyValueRegistration(BaseRegistration):
+    pass
 
 
 @json_schema_type
-class KeywordMemoryBankDef(CommonDef):
-    type: Literal[MemoryBankType.keyword.value] = MemoryBankType.keyword.value
+class KeywordRegistration(BaseRegistration):
+    pass
 
 
 @json_schema_type
-class GraphMemoryBankDef(CommonDef):
-    type: Literal[MemoryBankType.graph.value] = MemoryBankType.graph.value
-
-
-MemoryBankDef = Annotated[
-    Union[
-        VectorMemoryBankDef,
-        KeyValueMemoryBankDef,
-        KeywordMemoryBankDef,
-        GraphMemoryBankDef,
-    ],
-    Field(discriminator="type"),
+class GraphRegistration(BaseRegistration):
+    pass
+
+
+RegistrationRequest = Union[
+    VectorRegistration,
+    KeyValueRegistration,
+    KeywordRegistration,
+    GraphRegistration,
 ]
 
-MemoryBankDefWithProvider = MemoryBankDef
+
+def registration_request_to_memory_bank(request: RegistrationRequest) -> MemoryBank:
```

Comment:
we should put this in `providers/utils/memory` or somewhere in `distribution/` -- the `api` files should only have datatypes not functions wherever possible

Reply:
done.